Doc/issue220 create concepts file for parsers (#222)
DHBW-FN/OxideWM: concepts/config_files.md

> # Config files

Reviewer: ```suggestion
Window Manager Configuration
```

At the beginning of the project, it was decided to have an external config file which can be personalised freely.
For this to happen a suitable file format had to be chosen.

## File format

Typical file formats used for config files are `JSON`, `YAML` or `XML`.
`XML` was decided against because of readability.

> Compared to `JSON`, comments can be used in `YAML`, so the user can be guided trough the config file and does not need to study all variables and their function

Reviewer: ```suggestion
Unfortunately does `JSON` not support any comments inside the file, which was decided to be an important feature.
```

by hard.

> Therfore it was decided to use `YAML` as the proper file format for Oxide config files.

Reviewer 2: ```suggestion
Therefore it was decided to use `YAML` as the proper file format for Oxide config files.
```

## Technical implementation

> The following sections describe the technical solutions to enable the use of config files for Oxide.

Reviewer 2: ```suggestion
The following sections describe the technical solutions to enable the use of config files for *Oxide*.
```

### Parsing the config file

In order to be able to parse the values of the config file, it has to be serialized and deserialized.
To achive this, `serde_yaml` can be used. The `serde_yaml` crate is a rust library and therfore an optimal solution for the parser.

### Default values

> If the user does not create his own config file, it was necessary to have some default values in order for Oxide to work. `serde_yaml` implements the following

Reviewer 2: ```suggestion
If the user does not create his own config file, it was necessary to have some default values in order for *Oxide* to work. `serde_yaml` implements the following
```

function.

```rust
#[serde(default='default_value')]
```

Using this it is possible to use default values for not present or wrongly assigned variables.

## Conclusion

> After evaluating all aspects the team came to the conclusion to use `YAML` for the config file formar and `serde_yaml` to parse it.

Reviewer 2: ```suggestion
After evaluating all aspects the team came to the conclusion to use `YAML` for the config file format and `serde_yaml` to parse it.
```

> After evaluating all aspects the team came to the conclusion to use `YAML` for the config file formar and `serde_yaml` to parse it.

Reviewer: ### Parsing the config file
The config file needs to be parsed before we can accesses the stored data.
This should be as easy and effortless as possible. The preferred solution for this is to have a parser that outputs a single struct which contains all config values. 

### Library

The `serde` crate is the obvious choice for serialization and deserialization in the rust eco-system. It is widely supported and has subcrates such as `serde_yaml` for specific file formats. 

Additionally features such as giving fiels default values when not part of the config are possible.

## Conclusion

After evaluating all aspects the team came to the conclusion to use `YAML` as file format and the `serde_yaml` crate as parser.
